openedx/core/djangoapps/course_groups/cohorts.py

```diff
@@ -453,36 +448,41 @@ def add_user_to_cohort(cohort, username_or_email):
         User.DoesNotExist if a user could not be found.
     """
     try:
-        user = get_user_by_username_or_email(username_or_email)
+        if hasattr(username_or_email_or_user, 'email'):
+            user = username_or_email_or_user
+        else:
+            user = get_user_by_username_or_email(username_or_email_or_user)
 
-        membership = CohortMembership(course_user_group=cohort, user=user)
-        membership.save()  # This will handle both cases, creation and updating, of a CohortMembership for this user.
-        COHORT_MEMBERSHIP_UPDATED.send(sender=None, user=user, course_key=membership.course_id)
+        membership, previous_cohort = CohortMembership.assign(cohort, user)
         tracker.emit(
             "edx.cohort.user_add_requested",
             {
                 "user_id": user.id,
                 "cohort_id": cohort.id,
                 "cohort_name": cohort.name,
-                "previous_cohort_id": membership.previous_cohort_id,
-                "previous_cohort_name": membership.previous_cohort_name,
+                "previous_cohort_id": getattr(previous_cohort, 'id', None),
+                "previous_cohort_name": getattr(previous_cohort, 'name', None),
             }
         )
-        return (user, membership.previous_cohort_name, False)
+        cache = RequestCache(COHORT_CACHE_NAMESPACE).data
+        cache_key = _cohort_cache_key(user.id, membership.course_id)
+        cache[cache_key] = membership.course_user_group
```

Comment:
Future: As a subsequent iteration, we can look into refactoring the caching code.  Right now, we're updating the cache in multiple places.